sql/memory.go

```diff
@@ -0,0 +1,189 @@
+	MaxMemory() uint64
+	UsedMemory() uint64
+}
+
+// ProcessMemory is a reporter for the memory used by the process and the
+// maximum amount of money allowed controlled by the MAX_MEMORY environment
+// variable.
+var ProcessMemory Reporter = new(processReporter)
+
+type processReporter struct{}
+
+func (processReporter) UsedMemory() uint64 {
+	var s runtime.MemStats
+	runtime.ReadMemStats(&s)
+	return s.HeapInuse + s.StackInuse
+}
+
+func (processReporter) MaxMemory() uint64 { return maxMemory }
+
+// HasAvailableMemory reports whether more memory is available to the program if
+// it hasn't reached the max memory limit.
+func HasAvailableMemory(r Reporter) bool {
+	maxMemory := r.MaxMemory()
+	if maxMemory == 0 {
+		return true
+	}
+
+	return r.UsedMemory() < maxMemory
+}
+
+// MemoryManager is in charge of keeping track and managing all the components that operate
+// in memory. There should only be one instance of a memory manager running at the
+// same time in each process.
+type MemoryManager struct {
+	mu       sync.RWMutex
+	reporter Reporter
+	caches   map[uint64]Disposable
+	token    uint64
+}
+
+// NewMemoryManager creates a new manager with the given memory reporter. If nil is given,
+// then the Process reporter will be used by default.
+func NewMemoryManager(r Reporter) *MemoryManager {
+	if r == nil {
+		r = ProcessMemory
+	}
+
+	return &MemoryManager{
+		reporter: r,
+		caches:   make(map[uint64]Disposable),
+	}
+}
+
+// HasAvailable reports whether the memory manager has any available memory.
+func (m *MemoryManager) HasAvailable() bool {
+	return HasAvailableMemory(m.reporter)
+}
+
+// DisposeFunc is a function to completely erase a cache and remove it from the manager.
+type DisposeFunc func()
+
+// NewLRUCache returns an empty LRU cache and a function to dispose it when it's
+// no longer needed.
+func (m *MemoryManager) NewLRUCache(size int) (KeyValueCache, DisposeFunc) {
+	c := newLRUCache(m, m.reporter, size)
+	pos := m.addCache(c)
+	return c, func() {
+		c.Dispose()
+		m.removeCache(pos)
+	}
+}
+
+// NewHistoryCache returns an empty history cache and a function to dispose it when it's
+// no longer needed.
+func (m *MemoryManager) NewHistoryCache() (KeyValueCache, DisposeFunc) {
+	c := newHistoryCache(m, m.reporter)
+	pos := m.addCache(c)
+	return c, func() {
+		c.Dispose()
+		m.removeCache(pos)
+	}
+}
+
+// NewRowsCache returns an empty rows cache and a function to dispose it when it's
+// no longer needed.
+func (m *MemoryManager) NewRowsCache() (RowsCache, DisposeFunc) {
+	c := newRowsCache(m, m.reporter)
+	pos := m.addCache(c)
+	return c, func() {
+		c.Dispose()
+		m.removeCache(pos)
+	}
+}
+
+func (m *MemoryManager) addCache(c Disposable) (pos uint64) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.token++
+	m.caches[m.token] = c
+	return m.token
+}
+
+func (m *MemoryManager) removeCache(pos uint64) {
+	m.mu.Lock()
+	delete(m.caches, pos)
+
+	if len(m.caches) == 0 {
+		m.token = 0
+	}
+
+	m.mu.Unlock()
```

Comment:
Why isn't this `Unlock` deferred at the beginning as in `addCache`?

Reply:
Yep, probably better being deferred